### PR TITLE
PSIMA-145:Automating cache creation using initial conditions

### PR DESCRIPTION
**Pull request by sourabhdalvi**
That to make few changes to the existing methods
- Changed `get_stage_cache_definition`  to return an empty cache vector instead of nothing. (It is easier to populate empty array than a nothing)
- Added `_add_initial_condition_caches` that loops over the initial condition keys and checks if a cache is passed by the use, if not creates it.
- Changed `test_simulation_build` test to check for cache instead of error thrown

### CODE REVIEW

File: src/core/simulation.jl [hunk 1]
```diff
@@ -110,6 +110,61 @@ function unregister_recorders!(internal::SimulationInternal)
     end
 end
 
+function _add_initial_condition_caches(
+    sim::SimulationInternal,
+    stage::Stage,
+    caches::Union{Nothing, Vector{<:AbstractCache}},
```
Comment (jd-lara) [inline src/core/simulation.jl:116]: Add a debug statement here is the caches are `nothing`

File: src/core/simulation.jl [hunk 2]
```diff
@@ -110,6 +110,61 @@ function unregister_recorders!(internal::SimulationInternal)
...
+    return
+end
+
+function _create_cache(ic_key::ICKey, caches::Union{Nothing, Vector{<:AbstractCache}})
+    return
+end
+
+function _create_cache(
+    ic_key::ICKey{TimeDurationON, T},
+    caches::Union{Nothing, Vector{<:AbstractCache}},
```
Comment (jd-lara) [inline src/core/simulation.jl:131]: why a nothing here?

File: src/core/simulation.jl [hunk 3]
```diff
@@ -110,6 +110,61 @@ function unregister_recorders!(internal::SimulationInternal)
...
+        push!(caches, cache)
+    end
+    return
+end
+
+function _create_cache(
+    ic_key::ICKey{EnergyLevel, T},
+    caches::Union{Nothing, Vector{<:AbstractCache}},
+) where {T <: PSY.Device}
+
```
Comment (jd-lara) [inline src/core/simulation.jl:159]: I don't see why the caches will take a nothing here

Comment (sourabhdalvi) [reply src/core/simulation.jl:159]: I should actually remove that, but the reason I initial added it there was this function returned `Nothing`
https://github.com/NREL-SIIP/PowerSimulations.jl/blob/6ffdfa826540d66837af12af5b011d3c12e1ee77/src/core/simulation.jl#L242-L254

File: src/core/simulation.jl (hunk 2 above)
Comment (sourabhdalvi) [reply src/core/simulation.jl:131]: Sorry I forgot to remove that

File: src/core/simulation.jl (hunk 1 above)
Comment (sourabhdalvi) [reply src/core/simulation.jl:116]: Since the nothing will be removed that won't be needed

**Comment by sourabhdalvi** [comment]
> I have some questions about the input constraints to the `_create_cache` functions

Hopefully the comment above clear thing up. Short story, I forgot to remove it.

**Pull request merged**